### PR TITLE
[RSDK-1192] Multi camera robustness to order of plug/unplug

### PR DESCRIPTION
**Pull request by kharijarrett**
The solution here was to add a camRegistry (map) that will remember the label (by-path USB string) and the corresponding video path. This way, regardless of disconnections and the order of reconnection, AS LONG AS YOU USE THE SAME USB PORT, we will be able to reconnect to your same camera. This is a draft right now because I need to do cleaning and linting but I'm going home for now.

### CODE REVIEW

File: components/camera/videosource/webcam.go [hunk 1]
```diff
@@ -291,7 +293,19 @@ func NewWebcamSource(ctx context.Context, attrs *WebcamAttrs, logger golog.Logge
...
+	if !ok { // not in the map
+		camRegistry[attrs.Path] = label
+	} else { // is in the map
+		label = foundCamLabel
+	}
+
+	//label := attrs.Path
 	if label == "" {
 		label = getLabelFromCamera(utils.UnwrapProxy(cam).(camera.Camera), logger)
 	}
```
Comment (bazile-clyde) [inline components/camera/videosource/webcam.go:311]: If the call to `getLabelFromCamera` on line 298 sets `label := ""` we don't need to do it again here.

Comment (kharijarrett) [reply components/camera/videosource/webcam.go]: This is done.

**Comment by kharijarrett** [comment]
So fun fact: This entire PR exists because we have to link (under the hood) the label with the video path. If we expose the unique id of the camera to the user via drop down (as we plan to soon), there's no need for the video_path to be seen since it's a lot less user friendly. It would also fix this multicam ordering issue and make this whole PR moot in the best way possible!

In other words, I would totally not be offended if this just sat on ice until we made the relevant changes to the discovery service. I'd only be sad if we did neither.

**Comment by bazile-clyde** [comment]
> So fun fact: This entire PR exists because we have to link (under the hood) the label with the video path. If we expose the unique id of the camera to the user via drop down (as we plan to soon), there's no need for the video_path to be seen since it's a lot less user friendly. It would also fix this multicam ordering issue and make this whole PR moot in the best way possible!
> 
> In other words, I would totally not be offended if this just sat on ice until we made the relevant changes to the discovery service. I'd only be sad if we did neither.

Agreed. I'll push for exposing the `/dev/v4l/by-id/*` paths to the user instead of the `/dev/video*` paths in the scope ([RSDK-1756](https://viam.atlassian.net/browse/RSDK-1756)) since the former is more user-friendly. This will also have the nice side-effect of pinning a `go.viam.com/rdk/components/camera` to a `/dev/v4l/by-id/*` path which is unique to each camera even cameras of the same model.

For context here are the paths for a Raspberry Pi attached to a Viam rover
```
pi@raspberrypi:~ $ ls -l /dev/v4l/by-id
total 0
lrwxrwxrwx 1 root root 12 Jan 20 19:17 usb-GENERAL_GENERAL_WEBCAM-video-index0 -> ../../video0
lrwxrwxrwx 1 root root 12 Jan 20 19:17 usb-GENERAL_GENERAL_WEBCAM-video-index1 -> ../../video1
pi@raspberrypi:~ $ ls -l /dev/v4l/by-path/
total 0
lrwxrwxrwx 1 root root 13 Jan 20 19:17 platform-bcm2835-codec-video-index0 -> ../../video12
lrwxrwxrwx 1 root root 13 Jan 20 19:17 platform-bcm2835-isp-video-index0 -> ../../video20
lrwxrwxrwx 1 root root 13 Jan 20 19:17 platform-bcm2835-isp-video-index1 -> ../../video21
lrwxrwxrwx 1 root root 13 Jan 20 19:17 platform-bcm2835-isp-video-index2 -> ../../video22
lrwxrwxrwx 1 root root 13 Jan 20 19:17 platform-bcm2835-isp-video-index3 -> ../../video23
lrwxrwxrwx 1 root root 12 Jan 20 19:17 platform-fd500000.pcie-pci-0000:01:00.0-usb-0:1.1:1.0-video-index0 -> ../../video0
lrwxrwxrwx 1 root root 12 Jan 20 19:17 platform-fd500000.pcie-pci-0000:01:00.0-usb-0:1.1:1.0-video-index1 -> ../../video1
lrwxrwxrwx 1 root root 13 Jan 20 19:17 platform-feb10000.codec-video-index0 -> ../../video19
pi@raspberrypi:~ $ ls -l /dev/video*
crw-rw---- 1 root video 81, 14 Jan 20 19:17 /dev/video0
crw-rw---- 1 root video 81, 15 Jan 20 19:17 /dev/video1
crw-rw---- 1 root video 81,  8 Jan 20 19:17 /dev/video10
crw-rw---- 1 root video 81,  9 Jan 20 19:17 /dev/video11
crw-rw---- 1 root video 81, 10 Jan 20 19:17 /dev/video12
crw-rw---- 1 root video 81,  0 Jan 20 19:17 /dev/video13
crw-rw---- 1 root video 81,  1 Jan 20 19:17 /dev/video14
crw-rw---- 1 root video 81,  2 Jan 20 19:17 /dev/video15
crw-rw---- 1 root video 81,  3 Jan 20 19:17 /dev/video16
crw-rw---- 1 root video 81, 11 Jan 20 19:17 /dev/video18
crw-rw---- 1 root video 81, 13 Jan 20 19:17 /dev/video19
crw-rw---- 1 root video 81,  4 Jan 20 19:17 /dev/video20
crw-rw---- 1 root video 81,  5 Jan 20 19:17 /dev/video21
crw-rw---- 1 root video 81,  6 Jan 20 19:17 /dev/video22
crw-rw---- 1 root video 81,  7 Jan 20 19:17 /dev/video23
crw-rw---- 1 root video 81, 12 Jan 20 19:17 /dev/video31
```

Here's another example of two cameras of the same model whose `/dev/v4l/by-id` stays constant even after their `/dev/video*` paths change.
```
pi@raspberrypi:~ $ ls -l /dev/v4l/by-id
total 0
lrwxrwxrwx 1 root root 12 Jan 20 20:39 usb-EMEET_HD_Webcam_eMeet_C950_A220914000210019-video-index0 -> ../../video4
lrwxrwxrwx 1 root root 12 Jan 20 20:39 usb-EMEET_HD_Webcam_eMeet_C950_A220914000210019-video-index1 -> ../../video5
lrwxrwxrwx 1 root root 12 Jan 20 20:39 usb-EMEET_HD_Webcam_eMeet_C950_SN0001-video-index0 -> ../../video2
lrwxrwxrwx 1 root root 12 Jan 20 20:39 usb-EMEET_HD_Webcam_eMeet_C950_SN0001-video-index1 -> ../../video3
pi@raspberrypi:~ $ ls -l /dev/v4l/by-id
total 0
lrwxrwxrwx 1 root root 12 Jan 20 20:40 usb-EMEET_HD_Webcam_eMeet_C950_A220914000210019-video-index0 -> ../../video0
lrwxrwxrwx 1 root root 12 Jan 20 20:40 usb-EMEET_HD_Webcam_eMeet_C950_A220914000210019-video-index1 -> ../../video1
lrwxrwxrwx 1 root root 12 Jan 20 20:40 usb-EMEET_HD_Webcam_eMeet_C950_SN0001-video-index0 -> ../../video2
lrwxrwxrwx 1 root root 12 Jan 20 20:40 usb-EMEET_HD_Webcam_eMeet_C950_SN0001-video-index1 -> ../../video3

```

**Comment by tessavitabile** [comment]
I think this is a nice fix to have in place until we have the discovery service changes. Removed myself from reviewers, since I probably won't review further iterations of this PR.

File: components/camera/videosource/webcam.go [hunk 2]
```diff
@@ -290,10 +293,14 @@ func NewWebcamSource(ctx context.Context, attrs *WebcamAttrs, logger golog.Logge
...
 
-	label := attrs.Path
-	if label == "" {
-		label = getLabelFromCamera(utils.UnwrapProxy(cam).(camera.Camera), logger)
+	// grab this camera's label from the registry if it's there
+	foundCamLabel, ok := camRegistry[attrs.Path]
+	if !ok {
+		camRegistry[attrs.Path] = label
+	} else {
+		label = foundCamLabel
```
Comment (bhaney) [inline components/camera/videosource/webcam.go:303]: Test this out on macOS, and see if the paths still work as expected for that platform.

**Review by bhaney — COMMENTED** [review]
You'll want to test this out on macOS to see if it works too.

Also, attrs.Path is used on the first line of NewWebcamSource in `findCamera`. Should getting the label be done before findCamera?

File: components/camera/videosource/webcam.go (hunk 2 above)
Comment (kharijarrett) [reply components/camera/videosource/webcam.go:303]: Kinda blocked on macOS testing by (https://viam.atlassian.net/browse/RSDK-1790). Maybe we just consider this a Linux fix for now?

**Comment by bhaney** [comment]
We've been lingering on this PR for a while, I think mostly because it will be superseded by the discovery service work, and also doesn't really work for macOS. I know @kharijarrett said he was fine with closing this PR and delaying the fix until discovery service. I'm all right with that.

**Pull request closed without merging**